src/cpp/session/modules/SessionPackages.R

```diff
@@ -1226,79 +772,692 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
+      message <- paste(message, paste("Warning:", msgs, collapse = "\n"), sep = "\n")
+   }
+   
+   result <- list(
+      valid = file.exists(packagesFile) && is.null(error),
+      error = message
+   )
+   
+   .rs.scalarListFromList(result)
+   
+})
+
+.rs.addJsonRpcHandler("get_secondary_repos", function(cran, custom)
+{
+   .rs.getSecondaryRepos(cran, custom)
+})
+
+
+# a vectorized function that takes any number of paths and aliases the home
+# directory in those paths (i.e. "/Users/bob/foo" => "~/foo"), leaving any 
+# paths outside the home directory untouched
+.rs.addFunction("createAliasedPath", function(path)
+{
+   homeDir <- path.expand("~/")
+   homePathIdx <- substr(path, 1L, nchar(homeDir)) == homeDir
+   homePaths <- path[homePathIdx]
+   homeSuffix <- substr(homePaths, nchar(homeDir), nchar(homePaths))
+   path[homePathIdx] <- paste0("~", homeSuffix)
+   path
+})
+
+# Some R commands called during packaging-related operations (such as untar)
+# delegate to the system tar binary specified in TAR. On OS X, R may set TAR to
+# /usr/bin/gnutar, which exists prior to Mavericks (10.9) but not in later
+# rleases of the OS. In the special case wherein the TAR environment variable
+# on OS X is set to a non-existent gnutar and there exists a tar at
+# /usr/bin/tar, tell R to use that binary instead.
+if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
+    identical(Sys.getenv("TAR"), "/usr/bin/gnutar") && 
+    !file.exists("/usr/bin/gnutar") &&
+    file.exists("/usr/bin/tar"))
+{
+   Sys.setenv(TAR = "/usr/bin/tar")
+}
+
+.rs.addFunction("beforePackageUnloaded", function(package)
+{
+   # force any promises associated with package's registered S3 methods --
+   # this is necessary as otherwise the lazy-load database may become corrupt
+   # if a new version of this package is later installed as the old promises
+   # will now have invalid pointers to the old lazy-load database.
+   #
+   # note that we iterate over all loaded packages here because loading a
+   # package might entail registering S3 methods in the namespace of the
+   # package owning the generic, which typically is a separate package
+   #
+   # https://bugs.r-project.org/bugzilla/show_bug.cgi?id=16644
+   # https://github.com/rstudio/rstudio/issues/8265
+   for (namespaceName in loadedNamespaces())
+   {
+      .rs.tryCatch({
+         ns <- asNamespace(namespaceName)
+         table <- ns[[".__S3MethodsTable__."]]
+         as.list(table)
+      })
+   }
+})
+
+.rs.addFunction("updatePackageEvents", function()
+{
+   reportPackageStatus <- function(attached)
+   {
+      function(pkgname, ...)
+      {
+         packagePath <- .rs.pathPackage(pkgname, quiet = TRUE)
+         packageStatus = list(
+            name     = I(pkgname),
+            library  = I(dirname(packagePath)),
+            path     = I(.rs.createAliasedPath(packagePath)),
+            attached = I(attached)
+         )
```

Comment:
We now include both `library` and `path` here, to simplify upstream code in GWT land.